### DISCUSSION TITLE
Next.js Link and Lenis are not starting me at the top of the page when I navigate to a new page.

### DISCUSSION THREAD

**Discussion by joenemec** [darkroomengineering/lenis#244] [Q&A] [answered]
Most of the time when I click on a Next.js Link while using Lenis, I have no problem, but I noticed recently that sometimes it will start me at the bottom of the next page or sometimes in the middle of the next page. I don't have scroll={false}. I've had to remove lenis from my website because I don't want it to occur when clients are on the site. I've seen issues with Next.js Link but having it always take you to the top while using Lenis, but I haven't seen what I'm describing. Can anyone point me in the right direction?

**Comment by umarnauman0** [comment]
Hi, I think we had a similar issue. I was using vite and lenis and in my prod page it always would start from the bottom. I still don't know why but I came across this video that solved the issue for me.

https://www.youtube.com/watch?v=8c5eMyJR9Js&t=140s&ab_channel=BrianDesign

**Comment by clementroche** [comment]
Are you using <ReactLenis/> on app level ? not on each page ?

If so you have to "reset" lenis on every page by using lenis.scrollTo(0, {immediate: true, force:true})

**Reply by waynenguyen1302** [reply]
I was having the same problem and it was killing me, I tried lenis.scrollTo(0) and it didn't work. Thank you so much!

**Comment by kgalka** [comment]
Hello @clementroche 
I am not starting new thread, as this is a similar issue, maybe you would be able to help. 

I am using Next.js 14 and have a separate SmoothScrolling component (code below) which I include at the bottom of every page.tsx

It works almost perfect. The issue I got is when I click on the menu item while the scroll is still moving. Then it goes to the new page in a wrong position (it doesn't scroll to top). And this is actually problematic, since the case may easily occur when I click on the footer menu quickly after getting to the bottom of the page.

Any advice how to modify it to make it perfect?

PS. Thanks for the great work. It's so much better than the alternatives.

```
'use client';

import Lenis from '@studio-freight/lenis';

import { usePathname } from 'next/navigation';
import { useEffect, useRef } from 'react';

export default function SmoothScrolling() {
  const lenis = useRef<Lenis | null>(null);
  const pathname = usePathname();

  useEffect(() => {
    if (lenis.current) lenis.current!.scrollTo(0, { immediate: true });
  }, [pathname, lenis]);

  useEffect(() => {
    const lenis = new Lenis();

    lenis.on('scroll', (e: any) => {
      // console.log(e);
    });

    function raf(time: number) {
      lenis.raf(time);
      requestAnimationFrame(raf);
    }

    requestAnimationFrame(raf);

    return () => {
      lenis.destroy();
    };
  }, []);

  return <></>;
}
```

**Comment by clementroche** [comment]
You're telling me that scrollTo(0, {imediate:true}) is not interrupting scrolling ? Are you using a touchpad ? Can you provide a reproduction link ?

**Reply by kgalka** [reply]
Hey @clementroche I shared it privately on LinkedIn with you, I can't publish client's staging link here. 

I actually think that my setup isn't correct, as (lenis.current) is never true on pathname change... any suggestion how to rebuild it?

Currently I'm including the SmoothScrolling template in each page.tsx

EDIT: I am using MacBook touchpad to scroll

```
return (
    <main>
      <SliceZone slices={page.data.slices} components={components} />
      <SmoothScrolling />
    </main>
  );
```

**Reply by clementroche** [reply]
in your SmoothScrolling component, lenis.current is never assigned

**Reply by clementroche** [reply]
```js

'use client';

import Lenis from '@studio-freight/lenis';

import { usePathname } from 'next/navigation';
import { useEffect, useRef } from 'react';

export default function SmoothScrolling() {
  const lenisRef = useRef<Lenis | null>(null);
  const pathname = usePathname();

  useEffect(() => {
    if (lenisRef .current) lenisRef.current?.scrollTo(0, { immediate: true });
  }, [pathname]);

  useEffect(() => {
    const lenis = new Lenis();
    lenisRef.current = lenis

    lenis.on('scroll', (e: any) => {
      // console.log(e);
    });

    function raf(time: number) {
      lenis.raf(time);
      requestAnimationFrame(raf);
    }

    requestAnimationFrame(raf);

    return () => {
      lenis.destroy();
    };
  }, []);

  return <></>;
}

```

**Reply by kgalka** [reply]
This doesn't fix the issue. 

I also tried putting it inside layout.tsx instead of pages, but this breaks the scroll entirely (it doesn't refresh on path change).

**Reply by iamlinkus** [reply]
You've probably got something messed up in your code as it should work in layout. But if you're totally sure your code is set up correctly, one thing you could try is pass the scroll prop of false to your <Link> component. Here's that prop in next.js docs: https://nextjs.org/docs/app/api-reference/components/link#scroll

**Reply by ryanritz6** [reply]
Perfect fix to my issue, thanks so much!

**Comment by Ahsan-khan41** [comment]
Hey guys, I solved this problem using this:

```
const SmoothScroll: React.FC<Props> = ({ children }) => {
  const pathname = usePathname();
  const searchParams = useSearchParams();

  const lenis = useLenis();

  useEffect(() => {
    lenis?.scrollTo(0, { immediate: true });
  }, [pathname, searchParams, lenis]);

  return (
    <ReactLenis root options={{ lerp: 0.05 }}>
      {children}
    </ReactLenis>
  );
};
```

and wrap this component to your children in main layout.tsx

This worked for me, I hope works for others too. ✌

**Comment by wzee1** [comment]
This helped me in Next.js. It uses Next.js' Link and it also prefetches.
The issue that the user is not on top of the page when they scrolled & clicked the link at the same time is fixed by this.
Just use the LinkLenis instead of Link in your project.
"""
"use client"

import { ReactNode } from "react"
import { useRouter } from "next/navigation"
import { useLenis } from "@studio-freight/react-lenis"

import Link from "next/link"

type Props = {
  href: string
  children: ReactNode
  className?: string
  target?: string
}

const LinkLenis = ({ href, children, className, target }: Props) => {
  const router = useRouter()
  const lenis = useLenis()

  const handleClick = (e: React.MouseEvent<HTMLAnchorElement>) => {
    e.preventDefault()
    lenis?.stop()
    router.prefetch(href)
    router.push(href)
    lenis?.start()
  }

  return (
    <Link href={href} onClick={handleClick} className={className} target={target}>
      {children}
    </Link>
  )
}

export default LinkLenis
"""

**Reply by feledori** [reply]
you are using a deprecated version of lenis with "@studio-freight/react-lenis". please install "lenis" and use Lenis from "lenis/react".

**Reply by wzee1** [reply]
> you are using a deprecated version of lenis with "@studio-freight/react-lenis". please install "lenis" and use Lenis from "lenis/react".

Thank you, I already did that!

**Comment by QingjiaTsang** [comment]
🙃It still has this problem now!! I use vite, react18 and tanstack router here.  The way of lenis?.scrollTo(0, { immediate: true }) works very unnatural because it will scroll to the top before navigating to the other pages which is really wield.

**Reply by clementroche** [reply]
I don't know how tanstack works but you should wait for the next page to be ready.

**Reply by wzee1** [reply]
> 🙃It still has this problem now!! I use vite, react18 and tanstack router here.  The way of lenis?.scrollTo(0, { immediate: true }) works very unnatural because it will scroll to the top before navigating to the other pages which is really wield.

I already posted a Nextjs component on the thread solving this issue.

Try using that, but change the Nextjs' Link component to a normal "a" tag and use tanstack's router to push the href to the stack.

You can also prefetch the content before pushing to the stack, in this way, you can improve performance.

If you need help, throw my Nextjs code into an AI chat and ask it to convert it to vite & tanstack.

Good luck!

**Reply by QingjiaTsang** [reply]
Yeah, eventually I made a custom component to wrap the tanstack router Link to make it, but a bit different from your implements because I found that as long as I set the target props as "_top" to the Link component, then everything works well.

```
import { Link, LinkComponentProps } from "@tanstack/react-router"

function LenisLink({ target, ...props }: LinkComponentProps) {
  return (
    <Link target={target ?? "_top"} {...props} />
  );
}

export default LenisLink;
```

**Reply by QingjiaTsang** [reply]
Update: I found that if it's added the target prop, then it has a side effect which is like refresh the whole page (that's why the scroll seems to reset to the top of the page) making all my react query cache of the target route empty. Too bad because it means whenever doing navigation it'll unnecessarily fetch data again.

@wzee1 Then I tried your solution above, it still acts like the same old unnatual way which means not working for me.

I don't know if there's any other workarounds.

**Reply by QingjiaTsang** [reply]
Update: for those using tanstack router, here's the good solution for you.

<img width="744" alt="image" src="https://github.com/user-attachments/assets/a51efa26-6b55-4d59-a586-ca325b488c2c" />

You can literally use ScrollRestoration from tanstack router, and it monitors and maintains the scroll positions for every routes through the pathname key for the scroll restoration when you navigate back to the route which also provides a good user experience. Intuitively, if you're not navigating to the old visited route, but instead, a new route, then the scroll will be at the top position there.

reference: https://tanstack.com/router/latest/docs/framework/react/guide/scroll-restoration#scroll-restoration

**Comment by fanelfaa** [comment]
```tsx
'use client'

import { useEffect } from 'react'
import { usePathname } from 'next/navigation'
import ReactLenis, { useLenis } from 'lenis/react'

export function SmoothScroll({ children }: { children: React.ReactNode }) {
  const pathname = usePathname()

  const lenis = useLenis()

  useEffect(() => {
    if (lenis) {
      lenis.stop()
      requestAnimationFrame(() => {
        lenis.start()
      })
    }
  }, [pathname, lenis])
  return <ReactLenis root>{children}</ReactLenis>
}
```

this one work for me on next js 15 and lenis 1.2.3

**Comment by sahil-techessentia** [comment]
I had the same problem in nextjs15 app router, that's becuase i'm using loading.js file on pages, just removed the loading.js file and now it works fine

**Comment by clementroche** [comment]
Hello what about this solution?
       
```js
        function onClick(event: PointerEvent | MouseEvent) {
            const path = event.composedPath()
            const pageLink = path.find(
                (node) =>
                    node instanceof HTMLAnchorElement &&
                    (node.getAttribute("href")?.startsWith("./") ||
                        node.getAttribute("href")?.startsWith("/"))
            ) as HTMLAnchorElement | undefined
            if (pageLink) {
                if (lenis) {
                    lenis.reset() // this will stop lenis intertia
                }
            }
        }

        window.addEventListener("click", onClick)
```

**Comment by clementroche** [comment, marked as answer]
you'll need `stopInertiaOnNavigate` option.
